### PR TITLE
Remove Slots and replace HasDV with a hash of most of TxWitnesses

### PR DESCRIPTION
**Pull request by WhatisRT**

### CODE REVIEW

File: goguen/formal-spec/transactions.tex [hunk 1]
```diff
@@ -264,7 +223,7 @@ \subsection{Goguen transactions}
   that may be used by all non-native scripts in the transaction.
   The budget is intended to be computed by executing the scripts off-chain prior to transaction submission.
   \item We include a hash of the the indexed redeemer structure
-    with accessor $\fun{rdmrsHash}$. The reason for including this hash is explained below.
+    with accessor $\fun{sdHash}$. The reason for including this hash is explained below.
```
Comment (polinavino) [inline goguen/formal-spec/transactions.tex]: unless i missed it, I dont think the reason for this is actually explained anywhere.

I would suggest making a section about "Ensuring witness integrity" or "Including optional witness data for communication purposes" , or something like that.

File: goguen/formal-spec/utxo.tex [hunk 2]
```diff
@@ -746,17 +697,26 @@ \subsection{Witnessing}
...
     & ~~~~~~\var{c} \in \cwitness{cert} \cap \AddrScr\} \\
       \cup & ~~\{ (\var{pid}, \var{pid}) \mid \var{pid} \in \supp~(\fun{forge}~\var{txb}) \} \\
     & \where \\
     & ~~~~~~~ \var{txb}~=~\txbody{tx}
+    \nextdef
+    & \hspace{-1cm}\fun{hashSD} \in \TxWitness \to \ScriptDataHash^? \\
+    & \hspace{-1cm}\fun{hashSD}~\var{txw}~=~
+      \begin{cases}
+        \Nothing & (\var{sc}, \var{dat}, \var{rdmr}) ~=~ (\emptyset, \emptyset, \emptyset) \\
+        \fun{hash}~(\var{sc}, \var{dat}, \var{rdmr}) & \text{otherwise}
```
Comment (polinavino) [inline goguen/formal-spec/utxo.tex:714]: do we need `txw = (sc, dat, rdmr)` here?

File: goguen/formal-spec/utxo.tex [hunk 3]
```diff
@@ -818,34 +775,18 @@ \subsection{Witnessing}
...
-      \forall \var{validator} \in \fun{txscripts}~{txw} \cap \ScriptMSig,\\
-      \fun{runMSigScript}~\var{validator}~\var{tx}\\~\\
-      \{ h \mid (\wcard, h) \in \fun{scriptsNeeded}~\var{utxo}~\var{tx}\} ~=~ \dom (\fun{indexedScripts}~{tx}) \\
-      \forall (\var{purp}, h) \in ~\fun{scriptsNeeded}~\var{utxo}~\var{tx}, ~h\mapsto s~\in~\fun{indexedScripts}~{tx},\\
-       s \in \ScriptPlutus~\Leftrightarrow \exists r, \var{purp} \mapsto r \in \fun{findRdmr}~{tx}
+      \forall \var{validator} \in \fun{txscripts}~{txw} \cap \ScriptNative,\\
+      \fun{runNativeScript}~\var{validator}~\var{tx}\\~\\
+      \forall (\var{purp}, h) \in ~\fun{scriptsNeeded}~\var{utxo}~\var{tx}, \exists s, (h\mapsto s~\in~\fun{indexedScripts}~{tx} ~\land\\
+       s \in \ScriptNonNative~\Leftrightarrow \exists r, \var{purp} \mapsto r \in \fun{findRdmr}~{tx})
       \\~\\
```
Comment (polinavino) [inline goguen/formal-spec/utxo.tex]: I am not 100% behind the idea that all outputs have an optional datum - I prefer two types of outputs I think.
But either way, we need to check in the UTXOW rule that : 

for each output `o` corresponding to an input of the transaction being processed
- `o` has a datum (or is of TxOutP type) <=> its address corresponds to a plutus script

File: goguen/formal-spec/transactions.tex (hunk 1 above)
Comment (WhatisRT) [reply goguen/formal-spec/transactions.tex]: Thanks, I missed that. The previous sentence is also wrong :)

File: goguen/formal-spec/utxo.tex (hunk 2 above)
Comment (polinavino) [inline goguen/formal-spec/utxo.tex:719]: TxWitness includes the signatures too - i think you want `(sigs, sc, dat, rdmr)` and ignore sigs in the hashing

Comment (WhatisRT) [reply goguen/formal-spec/utxo.tex:719]: Note that below there's a where clause that matches the witnesses against that, ignoring the signatures

File: goguen/formal-spec/utxo.tex (hunk 3 above)
Comment (WhatisRT) [reply goguen/formal-spec/utxo.tex]: There isn't a type of Plutus (or non-native) scripts addresses right now though, and as far as I'm aware, we have no way of distinguishing those kinds of addresses from each other. This check would also imply that any output that has a datum but whose address does not correspond to a script would become unspendable (whereas now it just requires supplying the datum, without using it for anything). So then we would also have to ensure that a transaction doesn't generate these outputs.

Comment (polinavino) [reply goguen/formal-spec/utxo.tex]: we need one implication direction at least :
o's address corresponds to a plutus script => o has a datum (or is of TxOutP type)

you are right about the other direction - makes things unecessarily unspendable. but we need to explain this in the text

File: goguen/formal-spec/utxo.tex [hunk 4]
```diff
@@ -707,6 +707,11 @@ \subsection{Witnessing}
     & \where \\
     & ~~~~~~~ \var{txb}~=~\txbody{tx}
     \nextdef
+    & \hspace{-1cm}\fun{checkNativeScriptData} \in \Tx \to \UTxO \to (\ScriptPurpose \times \ScriptHash) \to \Bool \\
+    & \hspace{-1cm}\fun{checkNativeScriptData}~\var{tx}~\var{utxo}~(\var{purp},\var{h})~=~ \exists s, h\mapsto s~\in~\fun{indexedScripts}~{tx}\\
```
Comment (polinavino) [inline goguen/formal-spec/utxo.tex:711]: did you mean `checkNonNativeScriptData`?

**Review by polinavino — APPROVED** [review]
Excellent work! All the needed changes are done

**Pull request merged**